source/fuzz/transformation_inline_function.cpp

```diff
@@ -0,0 +1,166 @@
+// Copyright (c) 2020 André Perez Maselco
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "source/fuzz/transformation_inline_function.h"
+
+#include "source/fuzz/fuzzer_util.h"
+#include "source/fuzz/instruction_descriptor.h"
+
+namespace spvtools {
+namespace fuzz {
+
+TransformationInlineFunction::TransformationInlineFunction(
+    const spvtools::fuzz::protobufs::TransformationInlineFunction& message)
+    : message_(message) {}
+
+TransformationInlineFunction::TransformationInlineFunction(
+    const std::map<uint32_t, uint32_t>& result_id_map,
+    uint32_t function_call_id) {
+  *message_.mutable_result_id_map() =
+      google::protobuf::Map<google::protobuf::uint32, google::protobuf::uint32>(
+          result_id_map.begin(), result_id_map.end());
+  message_.set_function_call_id(function_call_id);
+}
+
+bool TransformationInlineFunction::IsApplicable(
+    opt::IRContext* ir_context, const TransformationContext& /*unused*/) const {
+  // |function_call_instruction| must be defined and must be an OpFunctionCall
+  // instruction.
+  auto function_call_instruction =
+      ir_context->get_def_use_mgr()->GetDef(message_.function_call_id());
+  if (function_call_instruction == nullptr ||
+      function_call_instruction->opcode() != SpvOpFunctionCall) {
+    return false;
+  }
+
+  // The called function must not have an early return.
+  auto called_function = fuzzerutil::FindFunction(
+      ir_context, function_call_instruction->GetSingleWordInOperand(0));
+  if (called_function->HasEarlyReturn()) {
+    return false;
+  }
+
+  for (auto& pair : message_.result_id_map()) {
```

Comment:
This loop checks that every key to the map does correspond to an instruction in the function being inlined, and complains if some key is not in the function.

I believe it *does not* currently check whether every instruction in the function has some associated key.

We should turn this around:

- It doesn't matter if the map contains redundant entries, and in fact it might happen during shrinking so we want to allow it

- It *does* matter if the map doesn't contain enough entries - it has to have an entry for every result id in the function that we are inlining

Can you change things to reflect this?

Reply:
Done.